### PR TITLE
vendor: github.com/fluent/fluent-logger-golang v1.8.0

### LINKED ISSUE #40063: Container hangs when using fluentd logger
**Description**
Docker container hangs when using fluentd logger with fluentd-async-connect=true and unreachable (turned off) fluentd server.

**Steps to reproduce the issue:**
1. Create docker-compose.yml with any container and logger set to
```
logging:
      driver: fluentd
      options:
        fluentd-address: 10.147.49.79:24224
        fluentd-async-connect: 'true'
        fluentd-retry-wait: '1s'
        fluentd-max-retries: '30'
        tag: your_app
```
2. `docker-compose up --build -d`

**Describe the results you received:**
Docker container hangs, does not respond to `docker-compose down -f`, `docker stop`, `docker kill`. 

**Describe the results you expected:**
Container exits with error message.

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue happens every time the remote fluentd server is down or unreachable.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.2
 API version:       1.40
 Go version:        go1.12.8
 Git commit:        6a30dfc
 Built:             Thu Aug 29 05:28:55 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.8
  Git commit:       6a30dfc
  Built:            Thu Aug 29 05:27:34 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 30
  Running: 29
  Paused: 0
  Stopped: 1
 Images: 47
 Server Version: 19.03.2
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-957.27.2.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 64
 Total Memory: 125.3GiB
 Name: server55
 ID: SHMQ:YUKH:MZOS:EEYR:WZI6:JYVH:NPPJ:CEEQ:ILXV:6JNU:BVBG:2NTE
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical

### PR DESCRIPTION
**Pull request by akerouanton**
Fixes #40063.

**- What I did**

Update the fluent logger library to v1.8.0. Following commits/PRs were merged since last bump:

* [Add callback for error handling when using async](https://github.com/fluent/fluent-logger-golang/pull/97)
* [Fix panic when accessing unexported struct field](https://github.com/fluent/fluent-logger-golang/pull/99)
* [Properly stop logger during (re)connect failure](https://github.com/fluent/fluent-logger-golang/pull/82)
* [Support a TLS-enabled connection](https://github.com/fluent/fluent-logger-golang/pull/107)
* [Do not allow writing events after fluent.Close()](https://github.com/fluent/fluent-logger-golang/pull/105)

**- How to verify it**

Before this change, running the following commands would lead to docker hanging on the `kill` command:

```console
$ docker run --rm -it --name test --log-driver=fluentd --log-opt fluentd-address=fluentdhost:24224 --log-opt fluentd-async=true debian /bin/bash -c 'echo test; sleep infinity;'
# In another terminal:
$ docker kill test
```

**- Description for the changelog**

- Forcefully stop fluentd logger when a container stop and the logger is running in async mode.

### CODE REVIEW

File: daemon/logger/fluentd/fluentd.go
Comment (akerouanton) [reply daemon/logger/fluentd/fluentd.go:54]: Oops, indeed I forgot the prefix.

Since dockerd could hang when this flag is disabled and fluentd is down, I think it's not really safe to disable it (unless #35716 got merged and timeout handling is improved in fluent-logger-golang). IMHO most people want to keep this flag enabled, but some few users might want to never lost logs (that's why I made it configurable). However, AFAIK dockerd doesn't log when containers logs can't be transmitted to fluentd, although this could be fixed by leveraging fluent/fluent-logger-golang#97 (which is vendored in this PR).

From your maintainer's point of view, is it ok if logs are lost when fluentd is down? Would you like to log the container logs that can't be transmitted?

Comment (PettitWesley) [reply daemon/logger/fluentd/fluentd.go:54]: Hey I'm from the Fluent community... 

> is it ok if logs are lost when fluentd is down?

Is there any other option? If Fluentd/bit is down, then the logs have nowhere to go. You wouldn't want to keep buffering them for too long. Discarding them seems like the right behavior. Is there a way for users to configure how long it will buffer and keep trying to re-connect before it starts just dropping logs- that seems ideal?

> although this could be fixed by leveraging fluent/fluent-logger-golang#97 (which is vendored in this PR).

What would the user experience look like if we used this?

Comment (akerouanton) [reply daemon/logger/fluentd/fluentd.go:54]: > > although this could be fixed by leveraging fluent/fluent-logger-golang#97 (which is vendored in this PR).
> 
> What would the user experience look like if we used this?

It'd send logs to journald so it'd be possible to see them through `# journalctl -u docker.service`. 

> Is there any other option? If Fluentd/bit is down, then the logs have nowhere to go. You wouldn't want to keep buffering them for too long. Discarding them seems like the right behavior.

No, logs have nowhere to go and, currently, because of how the exponential back-off retry algorithm is configured, it might block the daemon forever and that's exactly what this PR try to avoid.

> Is there a way for users to configure how long it will buffer and keep trying to re-connect before it starts just dropping logs- that seems ideal?

Short story: yes, but not implemented in Docker, and no, as it's probably broken in sync mode. Long story: Docker integration of fluentd doesn't provide a parameter to set lib's parameter `WriteTimeout`. Anyway, I believe the `WriteTimeout` mechanism is broken in sync mode under some circumstances: eg. trying to write a new log will change the write deadline of older logs, so under some relatively "high" load, deadlines will be pushed back forever such that no deadlines will be reached. To be more explicit, if the write timeout is set to 2 minutes and at least one log is written every <2m00, then no logs will time out. See [this comment](https://github.com/fluent/fluent-logger-golang/pull/82#issuecomment-949409791) (this needs proper testing and reporting).

---

Seeing your comments, I believe the best is to remove this flag. If logging to journald logs that can't be sent to fluentd is a desirable feature, I should open a new issue/PR. But I'd prefer to receive some guidance from docker maintainers about that first (cc @thaJeztah).

Comment (PettitWesley) [reply daemon/logger/fluentd/fluentd.go]: > Seeing your comments, I believe the best is to remove this flag. If logging to journald logs that can't be sent to fluentd is a desirable feature, I should open a new issue/PR. But I'd prefer to receive some guidance from docker maintainers about that first

I agree with this. Sounds like no one would want this current option. I am curious if some sort of max configurable retry attempts is possible?

If not, an optional dump logs to the docker journald log stream sounds like a useful option (default false tho). I think in practice very few of my users would use that, since they probably have a ton of container logs wouldn't want to risk the docker logs with their volume. If outputting to an option file or even falling back to the json log driver was possible... I think that would be an ideal user experience. But I would not put any great urgency on implementing that. Fixing the current bug here is most critical.

Comment (PettitWesley) [reply daemon/logger/fluentd/fluentd.go]: **The following comment is kind of long and rambly and so I want to be clear that helping me understand is fully optional** ... I really appreciate the work you are doing here and trust you as a maintainer/owner of this code... I had meant to try to read your code and try to deeply understand what it was actually doing... but was always busy with other work. I'm not a maintainer here and I don't have power to block or approve this, so I want to say that helping me understand is fully optional. 
 
> Docker integration of fluentd doesn't provide a parameter to set lib's parameter WriteTimeout. Anyway, I believe the WriteTimeout mechanism is broken in sync mode under some circumstances: eg. trying to write a new log will change the write deadline of older logs, so under some relatively "high" load, deadlines will be pushed back forever such that no deadlines will be reached. To be more explicit, if the write timeout is set to 2 minutes and at least one log is written every <2m00, then no logs will time out. See this comment (this needs proper testing and reporting).
 
 
Sorry I don't follow this and the linked comment. To me, it feels like if there's a single connection, and that connection blocks for a while, it should be possible to have a global timeout on that connection. Or is that not how it works? There's no single connection, each batch of logs is sent by a unique goroutine that can't safely have context on what others are seeing? And you need to tell Docker to give up at some point, and there's no way to coordinate around that? There's no way to have a "max async connection retries"? Are the only solutions possible- give up quickly, or risk being blocked forever? 
 
So the "Fluent" tool that I maintain is "Fluent Bit". Which is a low level C program that has multi-threading too. At a high level, this is how it works:
 
Log Producer => Fluent Bit (thing that has multiple threads accepting logs, and multiple threads outputting logs) => destination
 
And basically the way it works is that on the input side you can have timeouts on connections, and on the output side you can have a configurable max retries. So that even though each thread has no global context, it won't retry its data more than N times. And as far as shutting things down go, you set it up so that you stop your producer and Fluent Bit together. And if something goes wrong, Fluent Bit can't block the log producer from being shut down.
 
In this case I think we have this diagram:
 
Container stdout/stderr => Fluentd log driver => Fluentd/Fluent Bit instance
 
I wonder if its the same style of engineering problem... and thus that the following should be possible:
1. Log driver issues can never stop the container from shutting down. Why does that happen in docker??
2. You handle the destination being down with a max retries per chunk/log event

**Comment by thaJeztah** [comment]
Haven't fully read up on all of the comments, but w.r.t. "sending logs to journald", I think that should definitely be separate from this, and at least not be the default (it's somewhat unexpected that container logs would start to leak to journald).

Note that current versions of docker also keep a local buffer for reading back logs if a remote logging driver is used that doesn't support reading back logs (https://docs.docker.com/config/containers/logging/dual-logging/). So there's already a buffer of the last XX logging messages, even if the remote logging driver isn't able to handle them.

**Comment by PettitWesley** [comment]
@akerouanton @thaJeztah Can we get this moving? Lots of folks are blocked on this bug. (In addition to being a Fluent person, I'm also a logging person in general at AWS...)

And again, feel free to disregard the lengthy comments I made... helping me understand this is not something I'd ask to have priority- releasing this bug fix should be the priority.

**Comment by akerouanton** [comment]
@PettitWesley I'll update this PR this evening 😉

**Comment by akerouanton** [comment]
I removed the flag `force-stop-async-send` (it's now automatically true if `fluentd-async-connect` is true) and I bumped fluent-logger-golang to v1.8.0, the latest version releases 3 days ago. This new version:

1. Fixes a bug related to API consistency when running it in sync mode in multiple threads (which is not the case of docker), see [Do not allow writing events after fluent.Close()](https://github.com/fluent/fluent-logger-golang/pull/105) ;
2. introduces support for TLS connections, see [TLS Support](https://github.com/fluent/fluent-logger-golang/pull/107). However this change only supports self-signed CA certs by passing `TlsInsecureSkipVerify` parameter. So I don't think it's mature enough to advertise it to Docker users.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by tonistiigi — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by PettitWesley** [comment]
@thaJeztah I see this was added to the 21.xx milestone... will this come out in a patch release soon? This is a bug fix. It should be come out in the next patch release ASAP.

**Comment by cpuguy83** [comment]
@PettitWesley Definitely understand this fixes some issues for the built-in fluent logger.
We did discuss this today but didn't really decide on what to do since it's not *just* a bug fix.

I recommend using a plugin with the updated fluent for the time being.

**Comment by PettitWesley** [comment]
@cpuguy83 Thanks for the quick response! Once you make a decision, if you can update here or on the original issue, we'd highly appreciate that! https://github.com/moby/moby/issues/40063

In the meantime, I will look into seeing how I can get a custom patch build of this out to AWS customers.

**Pull request merged**
